src/api/app/controllers/announcements_controller.rb

```diff
@@ -0,0 +1,84 @@
+# TODO: Please overwrite this comment with something explaining the controller target
+class AnnouncementsController < ApplicationController
+  before_action :set_announcement, only: [:show, :update, :destroy]
+  # Pundit authorization policies control
+  after_action :verify_authorized, except: :index
+  after_action :verify_policy_scoped, only: :index
+
+  # GET /announcements
+  def index
+    @announcements = policy_scope(Announcement.all)
+    xml = if @announcements.exists?
+            @announcements.to_xml(Announcement::DEFAULT_RENDER_PARAMS)
+          else
+            '<announcements></announcements>'
+          end
+    render xml: xml
+  end
+
+  # GET /announcements/1
+  def show
+    if @announcement.present?
+      authorize @announcement
+      render xml: @announcement.to_xml(Announcement::DEFAULT_RENDER_PARAMS)
+    else
+      render_error message: 'Unknown announcement',
+                   status: 404, errorcode: 'unknown_announcement'
+    end
+  end
+
+  # POST /announcements
+  def create
+    @announcement = Announcement.new(announcement_params)
+    authorize @announcement
+    if @announcement.save
+      render xml: @announcement.to_xml(Announcement::DEFAULT_RENDER_PARAMS)
+    else
+      render_error message: @announcement.errors.full_messages,
+                   status: 400, errorcode: 'invalid_announcement'
+    end
+  end
+
+  # PATCH/PUT /announcements/1
+  def update
+    authorize @announcement
+    if @announcement.update(announcement_params)
+      render_ok
+    else
+      render_error message: @announcement.errors.full_messages,
+                   status: 400, errorcode: 'invalid_announcement'
+    end
+  end
+
+  # DELETE /announcements/1
+  def destroy
+    authorize @announcement
+    @announcement.destroy
+    render_ok
+  end
+
+  #### Non CRUD actions
+
+  #### Non actions methods
+  # Use hide_action if they are not private
+
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_announcement
+    @announcement = Announcement.find(params[:id])
+  end
+
+  # Only allow a trusted parameter "white list" through.
+  def announcement_params
```

Comment:
Because '<content/>' would be converted to content: {} and passing that to `@announcement.content` would set the attribute to `'{}'`

Reply:
Not really? You just pass the empty hash then to update/create which would set the attributes to nil / empty which should catch then your validation if it shouldn't be allowed :thinking: You can also do an empty check in the filter if you don't want to allow ``''``

``
render_error if hash.empty?
``

Reply:
My first approach was to use XmlHash, but as mentioned above this would require to sanitize the hash afterwards, or using a different way of parsing this.

In case you want to try it out yourself:
```ruby
xml = <<~XML
<announcement>
  <title/>
  <content>Terms of Service</content>
</announcement>
XML

xml_hash = Xmlhash.parse(xml).with_indifferent_access
attributes = xml_hash.slice(:title, :content)

announcement = Announcement.first
announcement.update(attributes)

> announcement.title
=> "{}"
```

Reply:
Yeah only if ``<title/>`` hmm

But anyway, I don't like to use nokogiri and xpath here ...